src/identities/Provider.ts

```diff
@@ -95,10 +98,23 @@ abstract class Provider {
     } catch (e) {
       return;
     }
-    // TODO: Add node ID validation here?
     if (!schema.claimIdentityValidate(claim)) {
       return;
     }
+    // We want to validate the NodeId in the data
+    try {
+      validateSync((keyPath, value) => {
+        return matchSync(keyPath)(
+          [['nodeId'], () => validationUtils.parseNodeId(value)],
+          [['nodeId1'], () => validationUtils.parseNodeId(value)],
+          [['nodeId2'], () => validationUtils.parseNodeId(value)],
+          () => value,
+        );
+      }, claim);
+    } catch (e) {
+      if (!(e instanceof validationErrors.ErrorParse)) return;
+      throw e;
+    }
```

Comment:
This doesn't make sense to me.

The `claim` is __claimed__ to be a `Claim` data structure.

This would be:

```ts
type Claim = {
  payload: {
    hPrev: string | null; // Hash of the previous claim (null if first claim)
    seq: number; // Sequence number of the claim
    data: ClaimData; // Our custom payload data
    iat: number; // Timestamp (initialised at JWS field)
  };
  signatures: Record<NodeIdEncoded, SignatureData>; // Signee node ID -> claim signature
};
```

This doesn't have anything like `nodeId`, or `nodeId1` ... etc.

What is this method for?  It's for basically doing a JSON parse and schema validate check on the claim string that is coming from the identity provider.

The schema check cannot check for node id, so this is meant to be a second-line of defense check on the node id.

Ok makes sense, however I'm not sure if this `claim` value even has those patterns at all.

Reply:
Those node ID fields come from the `ClaimData` type within the `Claim`:

```ts

// Claims can currently only be a cryptolink to a node or identity
type ClaimData = ClaimLinkNode | ClaimLinkIdentity;

// Cryptolink (to either a node or an identity)
type ClaimLinkNode = {
  type: 'node';
  node1: NodeIdEncoded;
  node2: NodeIdEncoded;
};
type ClaimLinkIdentity = {
  type: 'identity';
  node: NodeIdEncoded;
  provider: ProviderId;
  identity: IdentityId;
};
```

Remember we perform an HTTP request to get claims from providers (i.e. no GRPC). This `parseClaim` function seemed to be the best "boundary" to perform node ID validation.

Reply:
The return type of this function is:

```
  public parseClaim(identityClaimData: string): Claim | undefined {
```

That doesn't align with the expectation that `claim` is a `ClaimData` as you say.

Reply:
No, you're correct - `claim` is a `Claim` type. But, the `ClaimData` should be part of the `Claim` itself. We want to validate the node ID/s from that nested part of the claim.

e.g. a claim might be something like:

```
{
  payload: {
    hPrev: ...,
    seq: ...,
    data: {
      type: 'identity',
      node: 'someNodeIdToBeValidated',
      provider: ...,
      identity: ...
    },
    iat: ...,
  },
  signatures: { ... },
}
```

So, should it be something like this?:

```ts
    try {
      validateSync((keyPath, value) => {
        return matchSync(keyPath)(
          [['nodeId'], () => validationUtils.parseNodeId(value)],
          [['nodeId1'], () => validationUtils.parseNodeId(value)],
          [['nodeId2'], () => validationUtils.parseNodeId(value)],
          () => value,
        );
      }, claim.payload.data); // only pass the ClaimData value here?
```

Reply:
Furthermore, we can simplify this too. We never expect to have a `nodeId1`/`nodeId2` field, because the claims on providers are **always** going to be claims from node to identity. i.e. with `ClaimData` of the type:

```ts
type ClaimLinkIdentity = {
  type: 'identity';
  node: NodeIdEncoded;
  provider: ProviderId;
  identity: IdentityId;
};
```

Reply:
So the keypath should be `['payload', 'data', 'nodeId']`. Use the full keypath in case we need to validate other parts of the `claim`.

Reply:
I believe it should be `['payload', 'data', 'node']` - based on the type's field names.

Reply:
I've fixed this.